x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/analysis/Analyzer.java

```diff
@@ -825,4 +838,118 @@ private BitSet gatherPreAnalysisMetrics(LogicalPlan plan, BitSet b) {
         plan.forEachDown(p -> FeatureMetric.set(p, b));
         return b;
     }
+
+    private static class ImplicitCasting extends ParameterizedAnalyzerRule<LogicalPlan, AnalyzerContext> {
+        @Override
+        protected LogicalPlan rule(LogicalPlan plan, AnalyzerContext context) {
+            return plan.transformExpressionsUp(ScalarFunction.class, e -> ImplicitCasting.cast(e, context.functionRegistry()));
+        }
+
+        private static Expression cast(ScalarFunction f, FunctionRegistry registry) {
+            if (f instanceof EsqlScalarFunction esf) {
+                return processScalarFunction(esf, registry);
+            }
+
+            if (f instanceof EsqlArithmeticOperation eao) {
+                return processArithmeticOperation(eao);
+            }
+
+            return f;
+        }
+
+        private static List<DataType> getTargetDataType(String name, FunctionRegistry registry) {
+            List<FunctionDefinition> defs = registry.listFunctions()
+                .stream()
+                .filter(f -> f.clazz().getSimpleName().equalsIgnoreCase(name))
```

Comment:
Could you call `resolveFunction`?

Reply:
I'll build a new map for resolved functions, with the key as class name, so that we don't need to iterate here. After a function is resolved, its functionName is set as its class name, instead of the original function name, so the existing map with the original function name as keys cannot be used easily any more.